---
title: "Microbial Speciation"
author: "Ford Fishman"
date: "`r format(Sys.time(), '%d %B, %Y')`"
header-includes:
   - \usepackage{array}
output: pdf_document
geometry: margin=2.54cm
---

# Setup Work Environment

```{r}
rm(list=ls()) # removes all objects in the given environment
wd <- "~/Documents/LennonLab/GitHub/MicroSpeciation"
data_file <- paste(wd, "/data/MicroSpeciation.txt", sep = "")
```


```{r}
getwd()
setwd(wd)

# Load packages
require("png")
require("grid")
```

# Load data
```{r}
data <- read.table(data_file, sep = "\t", header = T)
```

# Exponential approach
Diversification involves the splitting of a lineage into two and so on, so by definition is an exponential process which has been used by evolutionary biologist to model speciation. We use the exponential function $S_f = S_0 * e^{rt}$, where Sf = species richness at time final, So is species richnness at time of origin which equals 1, r is the rate of speciation, and t is the amount of time between origin and final

```{r}
time <- data$Appeared - data$Extinction
rate <- (log(data$Richness/1)/time) # rate is log richness per time, Divide by 1?
data <- data.frame(data, time, rate)
data <- data[order(rate),] 
micro <- data[ which(data$Taxon == 'Microbes'),] # interesting way to subset
micro.r <- micro[,"rate"] # grabs the rate r from microbes ()
cat("Microbial speciation rate =", round(micro.r, 3), "per million years") # Basically a print statement, rounding the microbial evolutionary rate


# To see if this approach is reasonable, it would be interesting to compare (correlation) 
# literature-reported estimates of diversification for plants and animals with estimates from 
# the exponential model. Perhaps use data from Scholl and Weins (2016)
```

# Comparison with substitution-based diversification estimates
We can define a speciation event based on the time it takes for 16S rRNA gene of a given length to accumulate enough substitutions so that there is 3 % divergence. We can that put the corresponding speciation rate back into the exponential and see what levels of richness are predicted. 

```{r}
length.g <- 1600 # length of gene
cutoff <- 0.03 # % divergence for new OTU
subs = length.g * cutoff # number of substitutions needed for new OTU

# Substitution rates (per million years) from Kuo and Ochman (2009):
low.sub <- 0.4
hi.sub <- 1.5

# Corresponding divergence time (million years):
div.t.ls <- subs / low.sub # divergence time w/ low substitution rate
div.t.hs <- subs / hi.sub # divergence time w/ high substitution rate

# Corresponding speciation rates (million years):
# Might be more readable as r_low and r_high
spec.ls <- 1 / div.t.ls
spec.hs <- 1 / div.t.hs

# inverse of divergence times
cat("Speciation rates range from", round(spec.ls, 3), "to", round(spec.hs, 3), "per million years") 

# Let's plug these speciation rates back into the exponential model:
mic.yrs <- 4000 # millions of years of microbial evolution
# mic.yrs <- micro$time # alternate, avoid code duplication
Sf.ls <- 1 * exp(spec.ls*mic.yrs)
Sf.hs <- 1 * exp(spec.hs*mic.yrs)

cat("\nSpeciation rates from literature overshoot global Sobs:", formatC(Sf.ls, format = "e", digits = 2), "-", formatC(Sf.hs, format = "e", digits = 2)) 

# What would substitution rate need to be to get 10^12 species with exponential model?
right.sub <- micro.r * 48 # 0.33

# What if substitution rate was a bit lower (0.3)? How sensitive?
Sf.v.low <- 1 * exp(((right.sub - 0.03) /subs) * mic.yrs)

Sf.v.low <- 1 * exp(((right.sub - 0.03)/subs) * mic.yrs)
formatC(Sf.v.low, format = "e", digits = 2)
```

# What about a logistic model?
In population biology, a logistic model modifies exponential growth rate based on a carrying capacity. We don't know what the upper limit for species richness on Earth is. Perhaps it's 10^12 or 10^16. It however, cannot be 10^54 as calculated above with fast substitution rates as this would exceed the presumed equilibrium abundance (N) of the biosphere which is thought to be ~10^30. In calculations below, we will solve for richness with the following equation $S_f = \frac{K * S_o * e^{rt}} {K + S_o (e^{rt -1})}$

```{r}
K <- 10^14
SF.log <- (K * 1 * exp(spec.hs * mic.yrs)) / (K + 1 * exp(spec.hs * mic.yrs -1))
```

# Figure
```{r}
kern <- density(data$rate) # take the rate data from all organisms and compute kernal density
kernel_file <- paste(wd, "/figures/speciation-kernel.png", sep = "")
png(filename = kernel_file,
        width = 800, height = 800, res = 96*2)

par(mar = c(4, 5, 1, 1))  

plot(kern, main = NA, xaxt = "n", yaxt = "n", cex.lab = 1.5, ylab = "",
     xlab = "", xlim = c(-0.05, 0.16), ylim = c(0,12), lwd = 2)

mtext("Density", side = 2, outer = TRUE, cex = 1.5, 
      line = -2, adj = 0.5)

mtext(expression('Speciation Rate (mya'^-1*')'), side = 1, outer = TRUE, 
      cex = 1.5, line = -1, adj = 0.5)

     axis(side = 1, lwd.ticks = 2, cex.axis = 1.5, las = 1,
        at = c(-0.05, 0.0, 0.05, 0.10, 0.15), labels = T)

     axis(side = 3, lwd.ticks = 2, cex.axis = 1.5, las = 1,
        at = c(-0.05, 0.0, 0.05, 0.10, 0.15), labels = F)

     axis(side = 2, lwd.ticks = 2, cex.axis = 1.5, las = 1,
        at = c(0, 4, 8, 12), labels = T)

    axis(side = 4, lwd.ticks = 2, cex.axis = 1.5, las = 1,
        at = c(0, 4, 8, 12), labels = F)

box(lwd=2)

arrows(0.006907755, 4, 0.006907755, 7, length = 0.075, lwd = 2, col = "Black")
text(0.02, 3.5, labels = "microbes", cex = 1)

# Close Plot Device
dev.off()
graphics.off()

# Show Plot
img <- readPNG(kernel_file)
grid.raster(img)
```

# Comparison of Exponential and Phylogenetic Estimates of Speciation
```{r}
data.speciation <- read.table(paste(wd, "/data/Sholl_Wiens_2016.txt", sep = ""), sep = "\t", header = T)

exp.rate <- (log(data.speciation$richness/1)/data.speciation$age)

plot(data.speciation$e.div.0, exp.rate)
points(data.speciation$e.div.5, exp.rate, col = "red")
points(data.speciation$e.div.9, exp.rate, col = "blue")

plot(exp.rate, data.speciation$e.div.0, col = "green", ylab = "", xlab = "")
points(exp.rate, data.speciation$e.div.5, col = "red")
points(exp.rate, data.speciation$e.div.9, col = "blue")
points(exp.rate, exp.rate, type = "l", col = "black")
mtext("speciation rate (S & W 2016)", side = 2, outer = TRUE, cex = 1.5, 
      line = -2, adj = 0.5)
mtext("speciation rate (exponential)", side = 1, outer = TRUE, cex = 1.5, 
      line = -2, adj = 0.5)
```

